File: leetcode/easy/66-plusOne.java
// first pass solution
class Solution {
    public int[] plusOne(int[] digits) {
        int carry = 1;
        
        for (int i = digits.length - 1; i >= 0; i--) {
            digits[i] += carry;
            if (digits[i] > 9) {
                carry = 1;
                digits[i] = 0;
            } else {
                carry = 0;
            }
        }
        if (carry == 0) return digits; 
        
        int[] output = new int[digits.length + 1]; 
        output[0] = 1;
        
        for (int i = 0; i < digits.length; i++) {
            output[i+1] = digits[i]; 
        }
        return output;
    }
}

File: leetcode/easy/66-plusOne.cpp
class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        bool carry = true;

        for (int i = digits.size() - 1; i >= 0; i--) {
            if (carry && digits[i] == 9) {
                digits[i] = 0;
            } else {
                digits[i]++;
                carry = false;
                break;
            }
        }
        if (carry) {
            int next = digits[0];
            digits[0] = 1;
            for (int i = 1; i < digits.size(); i++) {
                int temp = digits[i];
                digits[i] = next;
                next = temp;
            }
            digits.push_back(next);
        }
        return digits;
    }
};
